src/SquidMath.h

```diff
@@ -38,54 +38,90 @@ using EnableIfType = typename std::enable_if<B,T>::type;
+}
 
+/// \returns a non-overflowing sum of the two unsigned arguments (or nothing)
+template <typename S, typename T, EnableIfType<AllUnsigned<S,T>::value, int> = 0>
+Optional<S>
+IncreaseSum(const S s, const T t) {
     // this optimized implementation relies on unsigned overflows
-    static_assert(std::is_unsigned<T>::value, "the first Sum(a,b) argument is unsigned");
-    static_assert(std::is_unsigned<U>::value, "the second Sum(a,b) argument is unsigned");
-    const auto sum = a + b;
-    // when a+b overflows, the result becomes smaller than any operand
-    return (sum < a) ? Optional<T>() : Optional<T>(sum);
+    static_assert(std::is_unsigned<S>::value, "the first argument is unsigned");
+    static_assert(std::is_unsigned<T>::value, "the second argument is unsigned");
+    // For the sum overflow check below to work, we cannot restrict the sum
+    // type which, due to integral promotions, may exceed common_type<S,T>!
+    const auto sum = s + t;
+    // 1. when summation overflows, the result becomes smaller than any operand
+    // 2. the unknown (see above) "auto" type may hold more than S can hold
+    return (s <= sum && sum <= std::numeric_limits<S>::max()) ?
+           Optional<S>(sum) : Optional<S>();
 }
 
-/// \returns a non-overflowing sum of the two signed arguments (or nothing)
-template <typename T, typename U, EnableIfType<!AllUnsigned<T,U>::value, int> = 0>
-Optional<T> constexpr
-Sum(const T a, const U b) {
-    // Instead of computing the largest type dynamically, we simply go by T and
-    // reject cases like Sum(0, LLONG_MAX) that would overflow on return.
-    static_assert(sizeof(T) >= sizeof(U), "Sum() return type can fit its (signed) result");
-
-    // tests below avoid undefined behavior of signed under/overflows
-    return b >= 0 ?
-           ((a > std::numeric_limits<U>::max() - b) ? Optional<T>() : Optional<T>(a + b)):
-           ((a < std::numeric_limits<U>::min() - b) ? Optional<T>() : Optional<T>(a + b));
+/// \returns a non-overflowing sum of the two arguments (or nothing)
+/// \returns nothing if at least one of the arguments is negative
+/// at least one of the arguments is signed
+template <typename S, typename T, EnableIfType<!AllUnsigned<S,T>::value, int> = 0>
+Optional<S> constexpr
+IncreaseSum(const S s, const T t) {
+    return
+        // We could support a non-under/overflowing sum of negative numbers, but
+        // our callers use negative values specially (e.g., for do-not-use or
+        // do-not-limit settings) and are not supposed to do math with them.
+        (Less(s, 0) || Less(t, 0)) ? Optional<S>() :
+        // Avoids undefined behavior of signed under/overflows. When A is not T,
+        // a or t undergoes (safe) integral conversion in these expressions.
+        // Sum overflow condition: s + t > maxS or, here, maxS - s < t.
+        // If the sum exceeds maxT, integral conversions will use S, not T.
```

Comment:
This does not make sense to me. There is no use of T here regardless of maxT and sum values, everything is in S type.

Reply:
The type of `t` is T. The sum in question (i.e. `s + t`) adds together a variable of type S and a variable of type T. The resulting sum type (_before_ it is copied to our `Optional<S>` object!) will be either S or T -- we do not know the resulting type; only the compiler does. We need to make sure that the sum does not overflow and that the subsequent copying into `Optional<S>` does not overflow.

* The sum may exceed maxS, and we handle/check for that explicitly.
* The sum may exceed maxT. Why do we _not_ check for that? This comment documents why. If the sum exceeds maxT but not maxS[1], then the sum will be of type S, not T due to C++ integral conversion rules (and we do handle/check for maxS overflows).

[1] Why "but not maxS"? Because this sum is computed _only_ if it does not exceed maxS. Our ternary operator is built that way.